### PR TITLE
add ssim observer 

### PR DESCRIPTION
**Pull request by wdwzyyg**
The SSIMObserver tracks the structural similarity (SSIM) of the object phase and probe intensity using multi-scale SSIM. The metric is computed as the average SSIM over three pyramid downsampling levels and compares the current state with that from a user-specified number of iterations earlier.

Also, the PatienceObserver can observe obj_ssim, probe_ssim, and loss jointly and will trigger termination whichever reaches the patience limit first. 

**Example set up in the YAML file:**
` calc_ssim: {every: 10, after: 10}` # same type of flag as save options

` early_termination_loss: 10` # note that the previous naming 'early_termination' is disabled. 

` early_termination_obj_ssim: 40` 

` early_termination_probe_ssim: null` 

**Showcase of an experimental reconstruction using calc_ssim: {every: 10}**

Screenshot 2026-04-10 at 9 26 10 AM
Screenshot 2026-04-10 at 9 26 30 AM

(the actual iter counts are saved in progress.obj_ssim.iters. e.g. here the value of progress.obj_ssim.values at index 0 corresponds to iter 20 v.s. iter 10. )

### CODE REVIEW

File: phaser/engines/common/simulation.py
Comment (hexane360) [inline phaser/engines/common/simulation.py:1]: It looks like these changes were mistakenly reverted from the upstream.

File: phaser/engines/gradient/run.py
Comment (hexane360) [inline phaser/engines/gradient/run.py:1]: these too

File: phaser/hooks/solver.py
Comment (hexane360) [inline phaser/hooks/solver.py:1]: and here

File: phaser/utils/_jax_kernels.py
Comment (hexane360) [inline phaser/utils/_jax_kernels.py:1]: and here

File: phaser/utils/analysis.py [hunk 1]
```diff
@@ -294,4 +294,149 @@ def align_and_correlate(mat: NDArray[numpy.floating]) -> NDArray[numpy.floating]
...
+
+    if 'cupy' in xp_name:
+        from cupyx.scipy.ndimage import zoom
+        zoom_factors = tuple(s1 / s2 for s1, s2 in zip(target_shape, im.shape))
+        return zoom(im, zoom_factors, order=1)
+
+    # JAX
+    import jax.image
+    return jax.image.resize(im, target_shape, method='linear')
+
```
Comment (hexane360) [inline phaser/utils/analysis.py]: Have you tested this stuff returns the same results on each backend? `affine_transform` in `phaser.utils.image` is correct on all backends, you could just add a helper there for `zoom` which calls to `affine_transform`. This also isn't the preferred way to test for backends, you should only need the helper functions in `phaser.utils.num`

File: phaser/utils/analysis.py [hunk 2]
```diff
@@ -294,4 +294,149 @@ def align_and_correlate(mat: NDArray[numpy.floating]) -> NDArray[numpy.floating]
...
+        sl_end = [slice(None)] * arr.ndim
+        sl_end[axis] = slice(size, size + n)
+        sl_beg = [slice(None)] * arr.ndim
+        sl_beg[axis] = slice(0, n)
+        return (cs[tuple(sl_end)] - cs[tuple(sl_beg)]) / size
+
+    return _along_axis(_along_axis(im, -2), -1)
+
+
+def structural_similarity(
```
Comment (hexane360) [inline phaser/utils/analysis.py:344]: should probably call this something different from `structural_similarity`, multi-scale is not the standard SSIM. Maybe `multiscale_structural_similarity`?

File: phaser/utils/num.py
Comment (hexane360) [inline phaser/utils/num.py:1]: more overwritten changes

File: phaser/execute.py [hunk 3]
```diff
@@ -316,6 +331,8 @@ def initialize_reconstruction(
     if init_state.tilt is not None and plan.init.tilt is None:
         logging.info("Re-using tilt from initial state...")
         tilt = init_state.tilt
+        tilt = tilt.astype(dtype)
+
     elif tilt_hook is not None:
```
Comment (hexane360) [inline phaser/execute.py:336]: Why was this changed? I don't think scan and tilt should have to be `dtype`, double-precision is fine

File: phaser/observer.py [hunk 4]
```diff
@@ -169,44 +169,144 @@ def finish_recons(self, state: ReconsState):
...
+                xp.angle(state.object.data),
+                xp.angle(self._prev_state.object.data),
+            )
+            state.progress['obj_ssim'].iters.append(total_iter)
+            state.progress['obj_ssim'].values.append(ssim_o)
+
+            ssim_p = structural_similarity(
+                xp.abs(state.probe.data),
+                xp.abs(self._prev_state.probe.data),
+            )
```
Comment (hexane360) [inline phaser/observer.py]: should provide `data_range`. Also, what does your structural_similarity do with complex numbers? Should you do real and imaginary parts separately?

File: phaser/plan.py [hunk 5]
```diff
@@ -57,21 +57,7 @@ class EnginePlan(Dataclass, kw_only=True):
...
-    Slices to unroll during JIT compilation (JAX backend only).
-    Larger unrolling may be faster, at the expense of increased compilation time.
-
-    `True` or `0` unrolls all slices, `False` or `1` disables unrolling.
-    `10` should be a good default value.
-    """
+    buffer_n_groups: int = 2
 
     update_probe: FlagLike = True
     update_object: FlagLike = True
```
Comment (hexane360) [inline phaser/plan.py:77]: more overwritten changes

File: phaser/state.py [hunk 6]
```diff
@@ -243,13 +243,6 @@ class PreparedRecons:
     name: str
     observer: 'ObserverSet'
 
-    def to_xp(self, xp: t.Any) -> Self:
-        return self.__class__(
-            self.patterns,
-            self.state.to_xp(xp),
-            self.name, self.observer,
-        )
-
```
Comment (hexane360) [inline phaser/state.py:252]: more overwritten changes

File: phaser/types.py [hunk 7]
```diff
@@ -387,12 +387,9 @@ def collect_errors(self, val: t.Any) -> t.Optional[ErrorNode]:
 
 
 __all__ = [
```
Comment (hexane360) [inline phaser/types.py:389]: more overwritten changes

File: pyproject.toml [hunk 8]
```diff
@@ -30,14 +30,15 @@ requires-python = ">=3.10"
...
+    "scikit-image>=0.19.0", 
     "matplotlib~=3.8",
     "h5py~=3.8",
     "pyyaml>=5.3.1",
     "click~=8.1.0",
     "rich>=12.0.0,<15",
     "tifffile>=2023.8.25",
     "optree>=0.13.0",
-    "py-pane==0.11.4",
+    "py-pane==0.11.3",
```
Comment (hexane360) [inline pyproject.toml]: overwritten change

File: pyproject.toml [hunk 9]
```diff
@@ -30,14 +30,15 @@ requires-python = ">=3.10"
 dependencies = [
     "numpy>=2.0,<2.6",  # tested on 2.3
     "scipy>=1.7.0,<1.19", # tested on 1.11, 1.16
+    "scikit-image>=0.19.0",
```
Comment (hexane360) [inline pyproject.toml]: is this needed? you don't seem to import it

File: phaser/observer.py [hunk 10]
```diff
@@ -169,44 +169,144 @@ def finish_recons(self, state: ReconsState):
...
+
+            logging.info(
+                f"SSIM (iters {prev_iter}→{total_iter}): "
+                f"obj={ssim_o:.4f}  probe={ssim_p:.4f}"
+            )
 
-        if self.no_improvement_iter >= self.patience:
-            logging.info(f"Early termination: no improvement for {self.patience} iterations")
-            raise EarlyTermination(state, self.continue_next_engine)
+        self._prev_state = deepcopy(state)
```
Comment (hexane360) [inline phaser/observer.py]: should take this off the GPU as well. Really you can take off the GPU for SSIM calculation there's no reason this has to be super fast

**Review by hexane360 — CHANGES_REQUESTED** [review]
It looks like your rebase didn't really work, you overwrote many changes to the develop branch. I made a couple other comments.

The structure of the changes seem mostly fine. The naming does seem pretty confusing. First is this is multi-scale SSIM not regular SSIM. Secondly, the relative to past iterations is potentially confusing, and it'll make it hard to add in relative to ground truth measurements later. Would names like `'obj_rel_msssim'` and `'probe_rel_msssim'` be less ambiguous?

File: phaser/execute.py (hunk 3 above)
Comment (wdwzyyg) [reply phaser/execute.py:336]: I've run into errors of dtype conflict when loading init state .h5 before, i am not sure if the recent changes have fixed that.

File: phaser/utils/analysis.py [hunk 11]
```diff
@@ -294,4 +294,139 @@ def align_and_correlate(mat: NDArray[numpy.floating]) -> NDArray[numpy.floating]
...
+        Box filter size in pixels (default 3).
+    num_scales : int
+        Number of pyramid levels (default 3).
+
+    Returns
+    -------
+    mssim : float
+        MS-SSIM value in [0, 1].
+    """
+    from phaser.utils.image import affine_transform as _affine_transform
```
Comment (hexane360) [inline phaser/utils/analysis.py:382]: ?

File: phaser/utils/analysis.py [hunk 12]
```diff
@@ -294,4 +294,139 @@ def align_and_correlate(mat: NDArray[numpy.floating]) -> NDArray[numpy.floating]
...
+
+    def _resample(im, target_shape):
+        scale_y = im.shape[-2] / target_shape[-2]
+        scale_x = im.shape[-1] / target_shape[-1]
+        matrix = numpy.array([[scale_y, 0.0], [0.0, scale_x]])
+        offset = numpy.array([0.5 * (scale_y - 1.0), 0.5 * (scale_x - 1.0)])
+        return _affine_transform(im, matrix, offset=offset, output_shape=target_shape[-2:], order=1)
```
Comment (hexane360) [inline phaser/utils/analysis.py:389]: did you test this thoroughly? it's easy to get off by 1 or off by 1/2 pixel errors here

File: phaser/utils/analysis.py [hunk 13]
```diff
@@ -294,4 +294,139 @@ def align_and_correlate(mat: NDArray[numpy.floating]) -> NDArray[numpy.floating]
...
+
+    if xp_name == 'numpy':
+        from scipy.ndimage import uniform_filter
+        return uniform_filter(im, sizes)
+
+    if 'cupy' in xp_name:
+        from cupyx.scipy.ndimage import uniform_filter
+        return uniform_filter(im, sizes)
+
+    # JAX or other: cumsum box filter along axes -2 and -1 only (XLA-friendly)
```
Comment (hexane360) [inline phaser/utils/analysis.py:322]: This is still not the correct way to do dispatch.
The following should work:

```python
if xp_is_jax(xp):
    # jax implementation

# numpy/cupy implementation
scipy = get_scipy_module(im)
return scipy.ndimage.uniform_filter(im, sizes)
```

And you need to make sure that the implementation returns the exact same numeric result as the scipy implementation.

File: phaser/observer.py [hunk 14]
```diff
@@ -169,44 +169,143 @@ def finish_recons(self, state: ReconsState):
...
-        if self.smoothed_error is None:
-            self.smoothed_error = error
-        else:
-            self.smoothed_error = (1 - self.smoothing) * self.smoothed_error + self.smoothing * error
+        xp = get_array_module(state.object.data)
+        total_iter = int(state.iter.total_iter)
+
+        # transfer only the two arrays needed; forces GPU→CPU sync here
+        obj_now   = to_numpy(xp.angle(state.object.data))
+        probe_now = to_numpy(xp.abs(state.probe.data))
```
Comment (hexane360) [inline phaser/observer.py:290]: Do you want to be doing just the phase of object and amplitude of probe? Have you compared the approaches?

File: phaser/observer.py [hunk 15]
```diff
@@ -169,44 +169,143 @@ def finish_recons(self, state: ReconsState):
...
-    def _error_from_state(self, state: t.Union[ReconsState, PartialReconsState]) -> t.Optional[float]:
-        if state.progress is None or (progress := state.progress['total_loss']) is None or not len(progress.values):
-            return None
-        return progress.values[-1]
+            iters_without_improvement = current_iter - self._last_improvement_iter.get(key, current_iter)
+            if iters_without_improvement >= patience:
+                logging.info(
+                    f"Early termination: {key} no improvement for {iters_without_improvement} iterations"
+                )
+                raise EarlyTermination(state, self.continue_next_engine)
```
Comment (hexane360) [inline phaser/observer.py:249]: How should we handle the following case?
```yaml
early_termination_obj_rel_msssim: 5
calc_rel_msssim: {'every': 10}
```

Right now it looks like it'll basically ignore patience; maybe an alternate way to specify is that the patience number is the number of *evaluations*, not the number of *iterations*

**Review by hexane360 — CHANGES_REQUESTED** [review]

**Comment by hexane360** [comment]
Also, I had one other thought: Do we really want to early terminate when the SSIM doesn't change much, or when the SSIM goes above or below a given threshold? The former seems like taking the derivative of the derivative. For instance, if the SSIM is constant at 0.9, that means a constant change in the reconstruction (i.e. constant improvement). But the current implementation would have that terminate.
